doc/syslog/syslog-rate-limit-design.md

```diff
@@ -0,0 +1,376 @@
+
+- Container scope:
+  - Multi ASIC: https://github.com/Azure/sonic-buildimage/blob/master/files/image_config/rsyslog/rsyslog-container.conf.j2
+  - Single ASIC: https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-base/etc/rsyslog.conf
+- Host scope: https://github.com/Azure/sonic-buildimage/blob/master/files/image_config/rsyslog/rsyslog.conf.j2
+
+Currently, each container has hardcoded message rate limiting to avoid receiving flooded log messages:
+
+```
+$SystemLogRateLimitInterval 300
+$SystemLogRateLimitBurst 20000
+```
+There is no rate limiting configured on host side for now.
+
+The SystemLogRateLimitInterval determines the amount of time that is being measured for rate limiting. The SystemLogRateLimitBurst defines the amount of messages, that have to occur in the time limit of SystemLogRateLimitInterval, to trigger rate limiting. For example, SystemLogRateLimitInterval=300, SystemLogRateLimitBurst=20000, it means that if one daemon generate more than 20000 messages in 300 seconds, rsyslogd will start to drop messages after that(FIFO).
+
+This feature allows user to configure SystemLogRateLimitInterval and SystemLogRateLimitBurst for host, containers.
+
+### Requirements
+
+- Support syslog message rate limit configuration for host and containers
+- New CLI for rate limit configuration
+- Have rate limit configuration persistent
+- Default rate limit shall be applied if no configuration provided
+- Both multi ASIC and single ASIC platform shall use rsyslog-container.conf.j2 as template
+- APP extension shall be able to declare its capability of this feature
+- APP extension is responsible for reading/listening rate limit configuration from CONFIG DB and set rsyslog configuration accordingly.
+
+### Architecture Design
+
+rsyslog.conf.j2 and rsyslog-container.conf.j2 shall be extended to accept template variable for SystemLogRateLimitInterval and SystemLogRateLimitBurst. rsyslog.conf for single ASIC platform will be removed and replaced by rsyslog-container.conf.j2.
+
+Host side:
+
+![host_config_on_fly](/doc/syslog/images/host_rate_limit_config_flow.svg).
+
+Change syslog message rate limit while container is running:
+
+![container_config_on_fly](/doc/syslog/images/container_rate_limit_config_flow.svg).
+
+For app extension, CLI shall only put the configuration to CONFIG DB and let app extension itself to handle it. How app extension handle CONFIG DB change is out of the feature scope.
+
+Load syslog message rate limit configuration during startup is already covered by existing flow, no change of the flow is required:
+
+- Host: rsyslog-config service automatically loads rate limit configuration from CONFIG DB and apply it to rsyslog
```

Comment:
is rsyslog-config a new service? why not reuse hostcfgd?

Reply:
I believe, we can use hostcfgd or new daemon (in host-services) to handle SYSLOG configs specific to containers and restart the corresponding container by executing "docker exec -i <container-name> supervisorctl restart restart-rsyslogd".

Reply:
I prefer to have a new service "syslogcfgd" to handle it. But it is still under discussion. Will update once done.

Reply:
Please share the design points here for wider community visibility.